storage/pebble/blocks_provider.go

```diff
@@ -0,0 +1,149 @@
+package pebble
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-evm-gateway/models"
+	"github.com/onflow/flow-evm-gateway/storage"
+	evmTypes "github.com/onflow/flow-go/fvm/evm/types"
+	flowGo "github.com/onflow/flow-go/model/flow"
+	gethCommon "github.com/onflow/go-ethereum/common"
+	"github.com/onflow/go-ethereum/eth/tracers"
+)
+
+// TODO(m-Peter): Remove ResultsCollector when local state index comes in
+type ResultsCollector struct{}
+
+func (s *ResultsCollector) StorageRegisterUpdates() map[flowGo.RegisterID]flowGo.RegisterValue {
+	return map[flowGo.RegisterID]flowGo.RegisterValue{}
+}
+
+var _ evmTypes.ReplayResultCollector = (*ResultsCollector)(nil)
+
+type blockSnapshot struct {
+	*BlocksProvider
+	block models.Block
+}
+
+var _ evmTypes.BlockSnapshot = (*blockSnapshot)(nil)
+
+func (bs *blockSnapshot) BlockContext() (evmTypes.BlockContext, error) {
+	return evmTypes.BlockContext{
+		ChainID:                evmTypes.EVMChainIDFromFlowChainID(bs.chainID),
+		BlockNumber:            bs.block.Height,
+		BlockTimestamp:         bs.block.Timestamp,
+		DirectCallBaseGasUsage: evmTypes.DefaultDirectCallBaseGasUsage,
+		DirectCallGasPrice:     evmTypes.DefaultDirectCallGasPrice,
+		GasFeeCollector:        evmTypes.CoinbaseAddress,
+		GetHashFunc: func(n uint64) gethCommon.Hash {
+			// For block heights greater than or equal to the current,
+			// return an empty block hash.
+			if n >= bs.block.Height {
+				return gethCommon.Hash{}
+			}
+			// If the given block height, is more than 256 blocks
+			// in the past, return an empty block hash.
+			if bs.block.Height-n > 256 {
+				return gethCommon.Hash{}
+			}
+
+			block, err := bs.blocks.GetByHeight(n)
+			if err != nil {
+				return gethCommon.Hash{}
+			}
+			blockHash, err := block.Hash()
+			if err != nil {
+				return gethCommon.Hash{}
+			}
+
+			return blockHash
+		},
+		Random: bs.block.PrevRandao,
+		Tracer: bs.tracer,
+	}, nil
+}
+
+type BlocksProvider struct {
+	blocks      storage.BlockIndexer
+	chainID     flowGo.ChainID
+	tracer      *tracers.Tracer
+	latestBlock *models.Block
+}
+
+var _ evmTypes.BlockSnapshotProvider = (*BlocksProvider)(nil)
+
+func NewBlocksProvider(
+	blocks storage.BlockIndexer,
+	chainID flowGo.ChainID,
+	tracer *tracers.Tracer,
+) *BlocksProvider {
+	return &BlocksProvider{
+		blocks:  blocks,
+		chainID: chainID,
+		tracer:  tracer,
+	}
+}
+
+func (bp *BlocksProvider) OnBlockReceived(block *models.Block) error {
+	if block == nil {
+		return fmt.Errorf("received nil block")
+	}
+
+	if bp.latestBlock != nil {
```

Comment:
Is it possible to eliminate this case by initializing BlocksProvider with a block?

Reply:
I've removed this in https://github.com/onflow/flow-evm-gateway/pull/632/commits/3cbf6ae8024c4d399f7d86d5a0f2840108fba573#diff-b3cc424323a936972f841745ae20117fcd028cecb7305a8ee4054023ca021e63R78-R90.

Basically I was thinking of `BlocksProvider` as a long-lived object, that is wired in the event ingestion component. 

Every time a new EVM block comes in (a.k.a latest block here, which is the latest received block from the Event Streaming API, that has not been replayed & indexed yet in EVM Gateway).

Right now, I only check that the received Block is sequential to the latest block, if there's any.